drivers/regulator/regulator_fixed.c

```diff
@@ -0,0 +1,392 @@
+	finalize_transition(data, notify, delay_us, rc);
+}
+#endif /* CONFIG_MULTITHREADING */
+
+static void start(struct onoff_manager *mgr,
+		  onoff_notify_fn notify)
+{
+	struct driver_data_onoff *data =
+		CONTAINER_OF(mgr, struct driver_data_onoff, mgr);
+	const struct driver_config *cfg = data->dev->config;
+	uint32_t delay_us = cfg->startup_delay_us;
+	int rc = 0;
+
+	LOG_DBG("%s: start", cfg->regulator_name);
+
+	if ((cfg->options & OPTION_ALWAYS_ON) != 0) {
+		delay_us = 0;
+		goto finalize;
+	}
+
+	rc = gpio_pin_set(data->gpio, cfg->gpio_pin, true);
+
+#ifdef CONFIG_MULTITHREADING
+	if (rc == -EWOULDBLOCK) {
+		/* Perform the enable and finalization in a work item.
+		 */
+		LOG_DBG("%s: start deferred", cfg->regulator_name);
+		__ASSERT_NO_MSG(data->task == WORK_TASK_UNDEFINED);
+		data->task = WORK_TASK_ENABLE;
+		data->notify = notify;
+		k_work_submit(&data->delayed_work.work);
+		return;
+	}
+#endif /* CONFIG_MULTITHREADING */
+
+finalize:
+	finalize_transition(data, notify, delay_us, rc);
+
+	return;
+}
+
+static void stop(struct onoff_manager *mgr,
+		 onoff_notify_fn notify)
+{
+	struct driver_data_onoff *data =
+		CONTAINER_OF(mgr, struct driver_data_onoff, mgr);
+	const struct driver_config *cfg = data->dev->config;
+	uint32_t delay_us = cfg->off_on_delay_us;
+	int rc = 0;
+
+	LOG_DBG("%s: stop", cfg->regulator_name);
+
+	if ((cfg->options & OPTION_ALWAYS_ON) != 0) {
+		delay_us = 0;
+		goto finalize;
+	}
+
+	rc = gpio_pin_set(data->gpio, cfg->gpio_pin, false);
+
+#ifdef CONFIG_MULTITHREADING
+	if (rc == -EWOULDBLOCK) {
+		/* Perform the disable and finalization in a work
+		 * item.
+		 */
+		LOG_DBG("%s: stop deferred", cfg->regulator_name);
+		__ASSERT_NO_MSG(data->task == WORK_TASK_UNDEFINED);
+		data->task = WORK_TASK_DISABLE;
+		data->notify = notify;
+		k_work_submit(&data->delayed_work.work);
+		return;
+	}
+#endif /* CONFIG_MULTITHREADING */
+
+finalize:
+	finalize_transition(data, notify, delay_us, rc);
+
+	return;
+}
+
+static int enable_onoff(const struct device *dev, struct onoff_client *cli)
+{
+	struct driver_data_onoff *data = dev->data;
+
+	return onoff_request(&data->mgr, cli);
+}
+
+static int disable_onoff(const struct device *dev)
+{
+	struct driver_data_onoff *data = dev->data;
+
+	return onoff_release(&data->mgr);
+}
+
+static const struct onoff_transitions transitions =
+	ONOFF_TRANSITIONS_INITIALIZER(start, stop, NULL);
+
+static const struct regulator_driver_api api_onoff = {
+	.enable = enable_onoff,
+	.disable = disable_onoff,
+};
+
+static int regulator_fixed_init_onoff(const struct device *dev)
+{
+	struct driver_data_onoff *data = dev->data;
+	int rc = common_init(dev, &data->gpio);
+
+	data->dev = dev;
+	onoff_manager_init(&data->mgr, &transitions);
+#ifdef CONFIG_MULTITHREADING
+	k_delayed_work_init(&data->delayed_work, onoff_worker);
+#endif /* CONFIG_MULTITHREADING */
+
+	if (rc >= 0) {
+		rc = 0;
+	}
+
+	LOG_INF("%s onoff: %d", dev->name, rc);
+
+	return rc;
+}
+
+struct driver_data_sync {
```

Comment:
struct comment?

Reply:
I suppose we can start requiring that for internal implementation data structures that aren't processed to provide documentation.  Do we want to do that?

Reply:
Yes I think we still want comments